File: C++/Grind75/Week-2/53.maximum-subarray.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution
{
public:
    int maxSubArray(vector<int> &nums)
    {
        int mxsum = INT_MIN, crsum = 0;
        for (auto &x : nums)
        {
            crsum += x;
            mxsum = max(mxsum, crsum);
            if (crsum < 0)
            {
                crsum = 0;
            }
        }
        return mxsum;
    }
};
// @lc code=end

File: Java/Grind75/Week2/53.maximum-subarray.java
package Grind75.Week2;
/*
 * @lc app=leetcode id=53 lang=java
 *
 * [53] Maximum Subarray
 */

// @lc code=start
class Solution {
    public int maxSubArray(int[] nums) {
        int currSum = 0, maxSum = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            currSum += nums[i];
            maxSum = Math.max(maxSum, currSum);
            if (currSum < 0)
                currSum = 0;
        }
        return maxSum;
    }
}
// @lc code=end
